### PR TITLE
📱 Improve Android Performance

### PR DESCRIPTION
**Pull request by grtcdr**
## What's changed
- There is a tiny little chance that CPU usage might work now with this patch using `ndk_sys::getloadavg()`, @uttarayan21 can you confirm this works?
- Implemented sysinfo for Android.
- expect() messages are now consistent with those of Linux & NetBSD.

### CODE REVIEW

File: src/android/mod.rs [hunk 1]
```diff
@@ -157,6 +158,27 @@ impl GeneralReadout for AndroidGeneralReadout {
         crate::shared::cpu_cores()
     }
 
+    fn cpu_usage(&self) -> Result<usize, ReadoutError> {
+        let nelem: i32 = 1;
+        let mut value: f64 = 0.0;
+        let value_ptr: *mut f64 = &mut value;
+        let cpu_load = unsafe { getloadavg(value_ptr, nelem) };
+        if cpu_load != -1 {
+            if let Ok(phys_cores) = cpu_cores() {
```
Comment (uttarayan21) [inline src/android/mod.rs:167]: This should be self.cpu_cores() else this will not compile

**Comment by uttarayan21** [comment]
Now it doesn't compile.
```bash
macchina on  main [!] is 📦 v0.7.0 via 🦀 v1.53.0-nightly took 3s 
❯ cross build --target=aarch64-linux-android --release
   Compiling libmacchina v0.2.5 (https://github.com/macchina-cli/libmacchina?branch=android-performance#3487c7f9)
   Compiling macchina v0.7.0 (/project)
error: linking with `aarch64-linux-android-gcc` failed: exit status: 1
  |
  = note: "aarch64-linux-android-gcc" "-Wl,--allow-multiple-definition" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-Wl,--as-needed" "-L" "/rust/lib/rustlib/aarch64-linux-android/lib" "/target/aarch64-linux-android/release/deps/macchina-b4bc4197cf27c9a1.macchina.86hkn7vu-cgu.0.rcgu.o" "-o" "/target/aarch64-linux-android/release/deps/macchina-b4bc4197cf27c9a1" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/target/aarch64-linux-android/release/deps" "-L" "/target/release/deps" "-L" "/target/aarch64-linux-android/release/build/jemalloc-sys-43e06b0bbb54bf95/out/build/lib" "-L" "/rust/lib/rustlib/aarch64-linux-android/lib" "-Wl,-Bstatic" "/tmp/rustcXx5oBN/libjemalloc_sys-1750bf7717a78456.rlib" "-Wl,--start-group" "-Wl,--end-group" "/rust/lib/rustlib/aarch64-linux-android/lib/libcompiler_builtins-be970cc3552dc34c.rlib" "-Wl,-Bdynamic" "-lgcc" "-landroid" "-ldl" "-llog" "-lgcc" "-lc" "-lm"
  = note: /target/aarch64-linux-android/release/deps/macchina-b4bc4197cf27c9a1.macchina.86hkn7vu-cgu.0.rcgu.o: In function `macchina::data::get_all_readouts::hac618efc115aaa70':
          macchina.86hkn7vu-cgu.0:(.text._ZN8macchina4data16get_all_readouts17hac618efc115aaa70E+0x4554): undefined reference to `getloadavg'
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

error: could not compile `macchina`

To learn more, run the command again with --verbose.
```

**Comment by grtcdr** [comment]
> Now it doesn't compile.
> 
> ```shell
> macchina on  main [!] is 📦 v0.7.0 via 🦀 v1.53.0-nightly took 3s 
> ❯ cross build --target=aarch64-linux-android --release
>    Compiling libmacchina v0.2.5 (https://github.com/macchina-cli/libmacchina?branch=android-performance#3487c7f9)
>    Compiling macchina v0.7.0 (/project)
> error: linking with `aarch64-linux-android-gcc` failed: exit status: 1
>   |
>   = note: "aarch64-linux-android-gcc" "-Wl,--allow-multiple-definition" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-Wl,--as-needed" "-L" "/rust/lib/rustlib/aarch64-linux-android/lib" "/target/aarch64-linux-android/release/deps/macchina-b4bc4197cf27c9a1.macchina.86hkn7vu-cgu.0.rcgu.o" "-o" "/target/aarch64-linux-android/release/deps/macchina-b4bc4197cf27c9a1" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/target/aarch64-linux-android/release/deps" "-L" "/target/release/deps" "-L" "/target/aarch64-linux-android/release/build/jemalloc-sys-43e06b0bbb54bf95/out/build/lib" "-L" "/rust/lib/rustlib/aarch64-linux-android/lib" "-Wl,-Bstatic" "/tmp/rustcXx5oBN/libjemalloc_sys-1750bf7717a78456.rlib" "-Wl,--start-group" "-Wl,--end-group" "/rust/lib/rustlib/aarch64-linux-android/lib/libcompiler_builtins-be970cc3552dc34c.rlib" "-Wl,-Bdynamic" "-lgcc" "-landroid" "-ldl" "-llog" "-lgcc" "-lc" "-lm"
>   = note: /target/aarch64-linux-android/release/deps/macchina-b4bc4197cf27c9a1.macchina.86hkn7vu-cgu.0.rcgu.o: In function `macchina::data::get_all_readouts::hac618efc115aaa70':
>           macchina.86hkn7vu-cgu.0:(.text._ZN8macchina4data16get_all_readouts17hac618efc115aaa70E+0x4554): undefined reference to `getloadavg'
>           collect2: error: ld returned 1 exit status
>           
> 
> error: aborting due to previous error
> 
> error: could not compile `macchina`
> 
> To learn more, run the command again with --verbose.
> ```

This is a linking issue...

**Comment by uttarayan21** [comment]
Same error is happening.

**Comment by grtcdr** [comment]
Compiles fine for me... Not sure why you're experiencing issues.

**Comment by uttarayan21** [comment]
Are you cross compiling for android ?

**Comment by grtcdr** [comment]
You're compiling macchina, not libmacchina. Does libmacchina compile just fine?

**Comment by grtcdr** [comment]
> Are you cross compiling for android ?
> If not you can use [this](https://github.com/rust-embedded/cross).

No I wasn't, I'll get it set up.

**Comment by uttarayan21** [comment]
Yes but compilation is not the problem is used incorrect terminology.

The problem is while linking the library with the android bionic library which doesn't seem to have a reference to the `loadavg` function.

**Comment by grtcdr** [comment]
> > You're compiling macchina, not libmacchina. Does libmacchina compile just fine?
> 
> Yes but compilation is not the problem is used incorrect terminology.
> 
> The problem is while linking the library with the android bionic library which doesn't seem to have a reference to the `loadavg` function.

Most likely a build.rs issue.

**Comment by grtcdr** [comment]
> > > You're compiling macchina, not libmacchina. Does libmacchina compile just fine?
> > 
> > 
> > Yes but compilation is not the problem is used incorrect terminology.
> > The problem is while linking the library with the android bionic library which doesn't seem to have a reference to the `loadavg` function.
> 
> ~Most likely a build.rs issue.~
> 
> I didn't read the docs, there seems to be more stuff that needs to be done in order to successfully call NDK functions, none of which I did, I'm not sure how to go about this, though.
> image

Scratch all that, we don't need ndk_sys, all we need is sysinfo, we can get the load average from there, which I already implemented for Linux.

**Comment by grtcdr** [comment]
We might be able to remove getprop calls by using this https://rust-lang.github.io/libc/aarch64-linux-android/doc/libc/fn.__system_property_get.html

**Comment by uttarayan21** [comment]
Even using sysinfo_ffi it doesn't seem to get the CPU usage.

Screenshot using the current build.
macchina

**Comment by uttarayan21** [comment]
Also this is the benchmark from this branch.

bench

**Comment by grtcdr** [comment]
> Also this is the benchmark from this branch.

Performance improvements aren't significant, the very important stuff hasn't been optimized yet.

**Pull request closed without merging**